parsers/vala.c

```diff
@@ -0,0 +1,570 @@
+
+	return lookupKeyword (s, lang);
+}
+
+static void readString (tokenInfo *token)
+{
+	int c;
+	while (1)
+	{
+		c = getcFromInputFile ();
+		switch (c)
+		{
+		case EOF:
+			return;
+		case '"':
+			tokenPutc (token, c);
+			return;
+		default:
+			tokenPutc (token, c);
+			break;
+		}
+	}
+}
+
+static void readIdentifier (tokenInfo *token)
+{
+
+	while (1)
+	{
+		int c = getcFromInputFile ();
+		if (c == EOF)
+			return;
+		else if ((!tokenIsEmpty (token)) &&
+				 (isalnum (c) || c == '_'))
+			tokenPutc (token, c);
+		else if (isalpha (c) || c == '_')
+			tokenPutc (token, c);
+		else
+		{
+			ungetcToInputFile (c);
+			break;
+		}
+	}
+}
+
+static void readToken (tokenInfo *const token, void *data CTAGS_ATTR_UNUSED)
+{
+	int c;
+
+	token->type		= TOKEN_UNDEFINED;
+	token->keyword	= KEYWORD_NONE;
+	vStringClear (token->string);
+
+ getNextChar:
+	/* Skip whitespaces */
+	do
+	{
+		c = getcFromInputFile ();
+	}
+	while (c == ' ' || c== '\t' || c == '\f');
+
+	if (c == '/')
+	{
+		int c0 = getcFromInputFile ();
+
+		if (c0 == '/')
+		{
+			/* line comment */
+			while ((c0 = getcFromInputFile ()))
+			{
+				if (c0 == EOF)
+				{
+					token->type = TOKEN_EOF;
+					return;
+				}
+				else if (c0 == '\n')
+					goto getNextChar;
+			}
+
+		}
+		else if (c0 == '*')
+		{
+			/* block comment */
+			int c1;
+			while ((c1 = getcFromInputFile ()))
+			{
+				if (c1 == EOF)
+				{
+					token->type = TOKEN_EOF;
+					return;
+				}
+				else if (c1 == '*')
+				{
+					int c2;
+				lookingForEndOfBlockComment:
+					c2 = getcFromInputFile ();
+					if (c2 == EOF)
+					{
+						token->type = TOKEN_EOF;
+						return;
+					}
+					else if (c2 == '/')
+						goto getNextChar;
+					else if (c2 == '*')
+						goto lookingForEndOfBlockComment;
+				}
+			}
+		}
+		else
+			ungetcToInputFile (c0);
+	}
+
+	switch (c)
+	{
+	case EOF:
+		token->type = TOKEN_EOF;
+		break;
+	case '"':
+		token->type = TOKEN_STRING;
+		tokenPutc (token, c);
+		readString (token);
+		break;
+	case '@':
+		token->type = TOKEN_IDENTIFIER;
+		tokenPutc (token, c);
+		readIdentifier (token);
+		break;
+	case '{':
+	case '}':
+	case '[':
+	case ']':
+	case '(':
+	case ')':
+	case '<':
+	case '>':
+	case ';':
+	case ',':
+		token->type = c;
+		tokenPutc (token, c);
+		break;
+	default:
+		if (isalpha (c) || c == '_')
+		{
+			tokenPutc (token, c);
+			readIdentifier (token);
+
+			token->keyword = resolveKeyword (token->string);
+			if (token->keyword == KEYWORD_NONE)
+				token->type = TOKEN_IDENTIFIER;
+			else
+				token->type = TOKEN_KEYWORD;
+			break;
+		}
+		else
+		{
+			token->type = c;
+			vStringPut (token->string, c);
+			break;
+		}
+	}
+}
+
+static tokenInfo *newValaToken (void)
+{
+	return newToken (&valaTokenInfoClass);
+}
+
+static void parseStatement (tokenInfo *const token)
+{
+	tokenInfo *lastToken = newValaToken ();
+	bool foundSignature = false;
+
+	do
+	{
+		tokenCopy (lastToken, token);
+		tokenRead (token);
+		if (tokenEqType (lastToken, 'class'))
```

Comment:
Is `'class'` a mistake of `'c'`?

Reply:
You are correct.

Reply:
Check please, I think it is fixed right now. Not sure if that is the supposed right approach to keep going.